### PR TITLE
Reduce branching when inserting components

### PR DESCRIPTION
**Pull request by james7132**
# Objective
We're currently using an unconditional `unwrap` in multiple locations when inserting bundles into an entity when we know it will never fail. This adds a large amount of extra branching that could be avoided on in release builds.

## Solution
Use `DebugCheckedUnwrap` in bundle insertion code where relevant. Add and update the safety comments to match.

This should remove the panicking branches from release builds, which has a significant impact on the generated code: https://github.com/james7132/bevy_asm_tests/compare/less-panicking-bundles#diff-e55a27cfb1615846ed3b6472f15a1aed66ed394d3d0739b3117f95cf90f46951R2086 shows about a 10% reduction in the number of generated instructions for `EntityMut::insert`, `EntityMut::remove`, `EntityMut::take`, and related functions.

### CODE REVIEW

File: crates/bevy_ecs/src/bundle.rs [hunk 1]
```diff
@@ -405,11 +409,11 @@ impl BundleInfo {
...
                 }
                 StorageType::SparseSet => {
-                    sparse_sets.get_mut(component_id).unwrap().insert(
-                        entity,
-                        component_ptr,
-                        change_tick,
-                    );
+                    let sparse_set =
+                        // SAFETY: If component_id is in self.component_ids, the caller guarentees that
+                        // a sparse set exists for the component.
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/bundle.rs:414]: These safety comments seem correct, but they doesn't really match the function's safety documentation:

> `table` must be the "new" table for `entity`. `table_row` must have space allocated for the `entity`, `bundle` must match this [`BundleInfo`]'s type

Can you change the function documentation to be more accurate? In particular it should mention how this applies to sparse set components. Seems like the callsites are all undocumented right now (oof) so it should be a simple change.

**Comment by james7132** [comment]
Definitely seems like there's a significant gain here. Microbenchmark results below:

<details>

```
group                                           main                                    reduced-branching-bundles
-----                                           ----                                    -------------------------
add_remove/sparse_set                           1.01   751.5±37.73µs        ? ?/sec     1.00   746.1±41.38µs        ? ?/sec
add_remove/table                                1.00  1122.5±18.47µs        ? ?/sec     1.00  1127.5±32.19µs        ? ?/sec
add_remove_big/sparse_set                       1.04  863.7±260.23µs        ? ?/sec     1.00  830.1±167.62µs        ? ?/sec
add_remove_big/table                            1.05      2.6±0.13ms        ? ?/sec     1.00      2.5±0.12ms        ? ?/sec
get_or_spawn/batched                            1.23   367.7±10.81µs        ? ?/sec     1.00   298.6±11.58µs        ? ?/sec
get_or_spawn/individual                         1.12   538.0±41.37µs        ? ?/sec     1.00   478.7±44.37µs        ? ?/sec
insert_commands/insert                          1.23   447.2±38.92µs        ? ?/sec     1.00   362.9±26.15µs        ? ?/sec
insert_commands/insert_batch                    1.33   385.9±16.20µs        ? ?/sec     1.00   290.3±17.74µs        ? ?/sec
insert_simple/base                              1.14    428.0±3.58µs        ? ?/sec     1.00    375.4±2.38µs        ? ?/sec
insert_simple/unbatched                         1.06   796.8±31.34µs        ? ?/sec     1.00   751.6±38.02µs        ? ?/sec
spawn_commands/2000_entities                    1.00    183.6±8.21µs        ? ?/sec     1.00    182.9±7.45µs        ? ?/sec
spawn_commands/4000_entities                    1.02   365.3±11.42µs        ? ?/sec     1.00    357.8±9.71µs        ? ?/sec
spawn_commands/6000_entities                    1.04   563.4±34.24µs        ? ?/sec     1.00   542.3±20.69µs        ? ?/sec
spawn_commands/8000_entities                    1.02   733.0±29.96µs        ? ?/sec     1.00   716.7±20.49µs        ? ?/sec
spawn_world/10000_entities                      1.00  841.9±117.49µs        ? ?/sec     1.10  929.8±168.96µs        ? ?/sec
spawn_world/1000_entities                       1.00    84.2±10.69µs        ? ?/sec     1.09    92.0±13.85µs        ? ?/sec
spawn_world/100_entities                        1.00      8.4±1.18µs        ? ?/sec     1.13      9.5±1.23µs        ? ?/sec
spawn_world/10_entities                         1.00  850.8±115.52ns        ? ?/sec     1.11  948.4±149.29ns        ? ?/sec
spawn_world/1_entities                          1.00    91.6±11.54ns        ? ?/sec     1.03    94.7±10.75ns        ? ?/sec
```
</details>

Separately noticed that there's still a few potential out-of-bounds panics when creating a BundleInserter or BundleSpawner that could be removed as well, but we can wait to do that in a separate PR.

**Comment by jdm** [comment]
It's interesting that all of the benchmarks are a bit faster, except for the spawn_world ones which are all consistently a bit slower.

**Comment by james7132** [comment]
> It's interesting that all of the benchmarks are a bit faster, except for the spawn_world ones which are all consistently a bit slower.

I'm not sure where this is coming from. The only change that might affect it in this PR is the changes to `write_components`.

File: crates/bevy_ecs/src/bundle.rs (hunk 1 above)
Comment (james7132) [reply crates/bevy_ecs/src/bundle.rs]: I moved the invariant to `BundleInfo::new` (moved from `initialize_bundle`), where there's only one callsite.

File: crates/bevy_ecs/src/bundle.rs [hunk 2]
```diff
@@ -266,6 +267,47 @@ pub struct BundleInfo {
 }
 
 impl BundleInfo {
+    /// Create a new [`BundleInfo`].
+    ///
+    /// # Safety
+    ///
+    /// `component_id` must be valid [`ComponentId`]'s in the same World and their associated storages must be
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/bundle.rs:274]:
```suggestion
    /// `component_ids` must be valid [`ComponentId`]'s belonging to the same World and their associated storages must be
```

File: crates/bevy_ecs/src/bundle.rs [hunk 3]
```diff
@@ -266,6 +267,47 @@ pub struct BundleInfo {
...
+            for id in component_ids {
+                if !seen.insert(id) {
+                    dups.push(id);
+                }
+            }
+
+            let names = dups
+                .into_iter()
+                .map(|id| {
+                    // SAFETY: component_id exists and is therefore valid
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/bundle.rs:299]:
```suggestion
                    // SAFETY: the caller ensures component_id is valid.
```

File: crates/bevy_ecs/src/bundle.rs [hunk 4]
```diff
@@ -393,7 +435,10 @@ impl BundleInfo {
             let component_id = *self.component_ids.get_unchecked(bundle_component);
             match storage_type {
                 StorageType::Table => {
-                    let column = table.get_column_mut(component_id).unwrap();
+                    let column =
+                        // SAFETY: If component_id is in self.component_ids, BundleInfo::new requires that
+                        // the target table contains the component.
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/bundle.rs:440]: It's probably be worth documenting this invariant on the `self.component_ids` field itself.

**Review by JoJoJet — APPROVED** [review]
Thank you for fixing the invariant. I think this PR is in a good state now, got a couple more suggestions though.

**Review by JoJoJet — APPROVED** [review]

**Pull request merged**